### PR TITLE
Squashed changes with YoloNASPose & Loss

### PR DESCRIPTION
**Pull request by BloodAxe**
This PR adds following components of YoloNASPose

* YoloNASPose model itself
* N/S/M/L variants and corresponding classes
* YoloNASPose loss and collate fn
* Exportable pose estimation interface

### CODE REVIEW

**Review by Louis-Dupont — COMMENTED** [review]
Here's some comments, but I'm still reviewing

File: src/super_gradients/module_interfaces/exportable_pose_estimation.py [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+            dtype_name = np.dtype(torch_dtype_to_numpy_dtype(input_image_dtype)).name
+            usage_instructions.append(
+                f"    example_input_image = np.zeros(({batch_size}, {input_image_channels}, {input_image_shape[0]}, {input_image_shape[1]})).astype(np.{dtype_name})"  # noqa
+            )
+
+            usage_instructions.append("    predictions = session.run(outputs, {inputs[0]: example_input_image})")
+            usage_instructions.append("")
+
+            usage_instructions.append("Exported model can also be used with TensorRT")
+            usage_instructions.append("To run inference with TensorRT, please see TensorRT deployment documentation")
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/module_interfaces/exportable_pose_estimation.py:564]: Aaactually I think this is better & easier to read :)

File: src/super_gradients/conversion/onnx/pose_nms.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+        """
+        batch_indexes, label_indexes, boxes_indexes = selected_indexes[:, 0], selected_indexes[:, 1], selected_indexes[:, 2]
+
+        selected_boxes = pred_boxes[batch_indexes, boxes_indexes]  # [num_detections, 4]
+        selected_scores = pred_scores[batch_indexes, boxes_indexes, label_indexes]  # [num_detections]
+        selected_poses = pred_joints[batch_indexes, boxes_indexes]  # [num_detections, Num Joints, 3]
+
+        # TODO: Refactor this
+        predictions = torch.cat([batch_indexes.unsqueeze(1), selected_boxes, selected_scores.unsqueeze(1), selected_poses.flatten(1)], dim=1)
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/conversion/onnx/pose_nms.py]: Nope, just leftover. Fixed in 8dc7647b1f901e34110398d9995429b49c8e7463

File: src/super_gradients/conversion/onnx/pose_nms.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+        pred_boxes = batch_predictions[:, :, 0:4]
+        pred_scores = batch_predictions[:, :, 4]
+        pred_joints = batch_predictions[:, :, 5:].reshape(self.batch_size, self.max_predictions_per_image, -1, 3)
+
+        return num_predictions.unsqueeze(1), pred_boxes, pred_scores, pred_joints
+
+    @classmethod
+    def as_graph(cls, batch_size: int, num_pre_nms_predictions: int, max_predictions_per_image, dtype: torch.dtype, device: torch.device) -> gs.Graph:
+        with tempfile.TemporaryDirectory() as tmpdirname:
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/conversion/onnx/pose_nms.py]: Fixed in 8dc7647b1f901e34110398d9995429b49c8e7463 thanks!

File: src/super_gradients/conversion/onnx/pose_nms.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+gs = import_onnx_graphsurgeon_or_fail_with_instructions()
+
+__all__ = ["attach_onnx_pose_nms"]
+
+
+class PoseNMSAndReturnAsBatchedResult(nn.Module):
+    __constants__ = ("batch_size", "max_predictions_per_image")
+
+    def __init__(self, batch_size: int, num_pre_nms_predictions: int, max_predictions_per_image: int):
+        super().__init__()
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/conversion/onnx/pose_nms.py]: Fixed in 8dc7647b1f901e34110398d9995429b49c8e7463 thanks!

File: src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        return decoded_predictions, raw_predictions
+
+    @property
+    def out_channels(self):
+        return None
+
+    def forward(self, feats: Tuple[Tensor]):
+        return self.forward_eval(feats)
+
+    def _generate_anchors(self, feats=None, dtype=None, device=None):
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py:196]: Fixed in 15cef48a8c9345355634119c633055a02dfa5432 thanks!

File: src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_variants.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+        return output_pred_bboxes, output_pred_scores, output_pred_joints
+
+
+class YoloNASPose(CustomizableDetector, ExportablePoseEstimationModel):
+    def __init__(
+        self,
+        backbone: Union[str, dict, HpmStruct, DictConfig],
+        heads: Union[str, dict, HpmStruct, DictConfig],
+        neck: Optional[Union[str, dict, HpmStruct, DictConfig]] = None,
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_variants.py:97]: Not doable. I think we cannot have backbone/heads/neck as instances here, since it is a CustomizableDetector who is actually instantiating stuff. Maybe it is not the most elegant solution as it limits us and forces using configs. But it is how it is.

File: src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    def _init_weights(self):
+        if self.eval_size:
+            device = infer_model_device(self)
+            dtype = infer_model_dtype(self)
+
+            anchor_points, stride_tensor = self._generate_anchors(dtype=dtype, device=device)
+            self.anchor_points = anchor_points
+            self.stride_tensor = stride_tensor
+
+    def forward_eval(self, feats: Tuple[Tensor, ...]):
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py]: Fixed in https://github.com/Deci-AI/super-gradients/commit/15cef48a8c9345355634119c633055a02dfa5432

File: src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+            head.replace_num_classes(num_classes, compute_new_weights_fn)
+
+        self.num_classes = num_classes
+
+    @staticmethod
+    def _pass_args(heads_list, factory, num_classes, reg_max):
+        for i in range(len(heads_list)):
+            heads_list[i] = factory.insert_module_param(heads_list[i], "num_classes", num_classes)
+            heads_list[i] = factory.insert_module_param(heads_list[i], "reg_max", reg_max)
+        return heads_list
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py]: Make sense to me. It was copy-pasted from Yolo-NAS without thinking too much :/
Fixed in 295b0c787cca4bb46232a9e45af801637c9b5fff

File: src/super_gradients/module_interfaces/exportable_pose_estimation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/module_interfaces/exportable_pose_estimation.py:564]: Not sure I understand. This is how I would use it
```python
# instructions.py
postprocessing_instructions = """Exported model contains postprocessing (NMS) step with the following parameters:"
num_pre_nms_predictions={num_pre_nms_predictions}"
max_predictions_per_image={max_predictions_per_image}"
nms_threshold={nms_threshold}"
confidence_threshold={confidence_threshold}"
output_predictions_format={output_predictions_format}"
"""
```

```python
# exportable_pose_estimation.py
from .instructions import postprocessing_instructions

...
if postprocessing:
    postprocessing_instructions.format(num_pre_nms_predictions=num_pre_nms_predictions, max_predictions_per_image=max_predictions_per_image, ...)
```
With all parameters hard coded.
Is this what you asked ? I'm not sure

File: src/super_gradients/training/datasets/pose_estimation_datasets/yolo_nas_pose_collate_fn.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def undo_flat_collate_tensors_with_batch_index(flat_tensor, batch_size: int) -> List[Tensor]:
+    """
+    Unrolls the flat tensor into list of tensors per batch item.
+    As name suggest it undoes what flat_collate_tensors_with_batch_index does.
+    :param flat_tensor:
+    :param batch_size:
+    :return: List of tensors
+    """
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/yolo_nas_pose_collate_fn.py]: Fixed in 408b5d0fcbf7a1c133cb659b19a8820df1342aa6

File: src/super_gradients/training/datasets/pose_estimation_datasets/yolo_nas_pose_collate_fn.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        is_crowd = flat_collate_tensors_with_batch_index(all_crowd_masks)
+        extras = {"gt_samples": batch}
+        return all_images, (boxes, joints, is_crowd), extras
+
+    def _generate_targets(self, sample: PoseEstimationSample) -> Tuple[Tensor, Tensor, Tensor]:
+        """
+        Generate targets for a single sample
+        :param sample:
+        :return:
+        """
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/yolo_nas_pose_collate_fn.py]: Fixed in 408b5d0fcbf7a1c133cb659b19a8820df1342aa6

File: src/super_gradients/training/losses/functional.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    cx = (b1_x1 + b1_x2) * 0.5
+    cy = (b1_y1 + b1_y2) * 0.5
+    return cx, cy
+
+
+def get_bbox_width_height(bbox: Tuple[Tensor, Tensor, Tensor, Tensor]):
+    b1_x1, b1_y1, b1_x2, b1_y2 = bbox
+    w = b1_x2 - b1_x1
+    h = b1_y2 - b1_y1
+    return w, h
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/losses/functional.py]: Fixed in e8b3f186bbce87674c81df209b04880d43e2460d

File: src/super_gradients/training/losses/functional.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+    area1 = (x2 - x1) * (y2 - y1)
+    area2 = (x2g - x1g) * (y2g - y1g)
+    union = area1 + area2 - overlap + eps
+    iou = overlap / union
+
+    return iou, overlap, union
+
+
+def get_convex_bbox(box1: Tuple[Tensor, Tensor, Tensor, Tensor], box2: Tuple[Tensor, Tensor, Tensor, Tensor]):
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/losses/functional.py]: Fixed in e8b3f186bbce87674c81df209b04880d43e2460d

File: src/super_gradients/training/losses/functional.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    :param box1:    box1 with the shape (..., 4)
+    :param box2:    box1 with the shape (..., 4)
+    :param eps:     epsilon to avoid divide by zero
+    :return:
+        - iou:      iou of box1 and box2
+        - overlap:  overlap of box1 and box2
+        - union:    union of box1 and box2
+    """
+    x1, y1, x2, y2 = box1
+    x1g, y1g, x2g, y2g = box2
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/losses/functional.py]: Fixed in e8b3f186bbce87674c81df209b04880d43e2460d

File: src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+from super_gradients.training.utils.utils import infer_model_dtype, infer_model_device
+
+
+@register_detection_module()
+class YoloNASPoseNDFLHeads(BaseDetectionModule, SupportsReplaceNumClasses):
+    def __init__(
+        self,
+        num_classes: int,
+        in_channels: Tuple[int, int, int],
+        heads_list: Union[str, HpmStruct, DictConfig],
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/pose_estimation_models/yolo_nas_pose/yolo_nas_pose_ndfl_heads.py]: I generally agree that we should allow passing already instantiated heads. 
However in current design of the entire model instantiation I don't see a good way to support this 🤷 
Perhaps, a right way to approach this would be an introduction of classmethods `from_config` to discriminate the intention:

```python
class YoloNASPoseNDFLHeads(BaseDetectionModule, SupportsReplaceNumClasses):
  def __init__(
    self,
    heads_list: List[YoloNASPoseHead]
  ):
    self.heads_list = heads_list
  
  @classmethod
  def from_config(cls, config):
    heads_list = INSTANTIATE HEADS FROM CONFIG
    return cls(heads_list)
```

But again, this is pretty massive refactor and I'm not sure how to approach this now

**Review by Louis-Dupont — APPROVED** [review]
LGTM
